File: src/Tree/PathSum/PathSum1.java
class Solution {
    public boolean hasPathSum(TreeNode root, int targetSum) {
        if(root == null) return false;
        return dfs(root, targetSum);
    }
    private boolean dfs(TreeNode root, int targetSum){
        targetSum -= root.val;
        //Base case: leaf node
        if(root.left == null && root.right == null){
            return targetSum == 0;
        }
        boolean left = false, right = false;
        if(root.left != null){
            left = dfs(root.left, targetSum);
        }
        if(root.right != null){
            right = dfs(root.right, targetSum);
        }
        return left || right;
    }
}

File: src/Tree/PathSum/PathSum1.py
class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        # Base case        
        if root is None:
            return False
        
        targetSum -= root.val
        
        if root.left is None and root.right is None:
            return targetSum == 0
        
        return self.hasPathSum(root.left, targetSum) or self.hasPathSum(root.right, targetSum)
